src/backend/distributed/commands/dependencies.c

```diff
@@ -182,14 +182,40 @@ ErrorIfCircularDependencyExists(const ObjectAddress *objectAddress)
 
 
 /*
- * DeferErrorIfCircularDependencyExists checks whether given object has
- * circular dependency with itself via existing objects of pg_dist_object.
+ * DeferErrorIfCircularDependencyExists is a wrapper function around
+ * DeferErrorIfCircularDependencyExistsOnObject, which calls that function for
+ * each dependency of the given object.
  */
 DeferredErrorMessage *
 DeferErrorIfCircularDependencyExists(const ObjectAddress *objectAddress)
 {
 	List *dependencies = GetAllSupportedDependenciesForObject(objectAddress);
 
+	ObjectAddress *dependency = NULL;
+	foreach_ptr(dependency, dependencies)
```

Comment:
I think we should see the performance impact of this change. For example, can you create some tests like: https://gist.github.com/onderkalaci/01b83abfc849995d406a292084391aa3

and see how much overhead this adds? 100k dependent objects could be a little extreme, but that happens.
I think adding overhead is ok, but that should be bearable

Reply:
Tested with 16k objects. Execution time increased from ~5.20 minutes to ~7.20 minutes.
(Though I think most of the overhead is because of the CREATE commands)

Reply:
I'm afraid my experience is different. I think we cannot backport such an invasive change. Can we make this a proper graph traversal instead of ~O(N!) change ?

 I use this very small dependency: https://gist.githubusercontent.com/onderkalaci/7513ee82ce843cbda374653610952304/raw/2a126426492c554ded3481c4241b73f81ffe691e/small_objects.sql


On main, it takes ~ 3 seconds to create all
```
time psql -f small_objects.sql postgres

real	0m3.945s
user	0m0.010s
sys	0m0.010s
```

And, on this branch, it takes ~100 seconds  with CPU jumping to %100 percent. 

```
 time psql -f small_objects.sql postgres
real	1m39.106s
user	0m0.012s
sys	0m0.012s
```